include/psa/crypto_extra.h

```diff
@@ -1904,37 +1909,100 @@ static inline void psa_pake_cs_set_hash( psa_pake_cipher_suite_t *cipher_suite,
+    PSA_PAKE_X2_STEP_KEY_SHARE   = 4,  /* Round 1: input/output key share (for ephemeral private key X2).*/
+    PSA_PAKE_X2_STEP_ZK_PUBLIC   = 5,  /* Round 1: input/output Schnorr NIZKP public key for the X2 key */
+    PSA_PAKE_X2_STEP_ZK_PROOF    = 6,  /* Round 1: input/output Schnorr NIZKP proof for the X2 key */
+    PSA_PAKE_X2S_STEP_KEY_SHARE  = 7,  /* Round 2: output X2S key (our key) */
+    PSA_PAKE_X2S_STEP_ZK_PUBLIC  = 8,  /* Round 2: output Schnorr NIZKP public key for the X2S key (our key) */
+    PSA_PAKE_X2S_STEP_ZK_PROOF   = 9,  /* Round 2: output Schnorr NIZKP proof for the X2S key (our key) */
+    PSA_PAKE_X4S_STEP_KEY_SHARE  = 10, /* Round 2: input X4S key (from peer) */
+    PSA_PAKE_X4S_STEP_ZK_PUBLIC  = 11, /* Round 2: input Schnorr NIZKP public key for the X4S key (from peer) */
+    PSA_PAKE_X4S_STEP_ZK_PROOF   = 12  /* Round 2: input Schnorr NIZKP proof for the X4S key (from peer) */
+};
+
+struct psa_jpake_computation_stage_s
 {
-    psa_algorithm_t MBEDTLS_PRIVATE(alg);
     unsigned int MBEDTLS_PRIVATE(state);
     unsigned int MBEDTLS_PRIVATE(sequence);
-#if defined(MBEDTLS_PSA_BUILTIN_PAKE)
     unsigned int MBEDTLS_PRIVATE(input_step);
     unsigned int MBEDTLS_PRIVATE(output_step);
-    uint8_t* MBEDTLS_PRIVATE(password);
-    size_t MBEDTLS_PRIVATE(password_len);
-    psa_pake_role_t MBEDTLS_PRIVATE(role);
-    uint8_t MBEDTLS_PRIVATE(buffer[MBEDTLS_PSA_PAKE_BUFFER_SIZE]);
-    size_t MBEDTLS_PRIVATE(buffer_length);
-    size_t MBEDTLS_PRIVATE(buffer_offset);
-#endif
-    union
-    {
-#if defined(MBEDTLS_PSA_BUILTIN_ALG_JPAKE)
-        mbedtls_ecjpake_context ecjpake;
-#endif
-        /* Make the union non-empty even with no supported algorithms. */
-        uint8_t dummy;
-    } MBEDTLS_PRIVATE(ctx);
+};
+
+struct psa_pake_computation_stage_s
+{
+    union {
+        unsigned dummy;
+        psa_jpake_computation_stage_t MBEDTLS_PRIVATE(jpake_computation_stage);
+    } MBEDTLS_PRIVATE(data);
+};
+
+struct psa_pake_operation_s
+{
+    /** Unique ID indicating which driver got assigned to do the
+     * operation. Since driver contexts are driver-specific, swapping
+     * drivers halfway through the operation is not supported.
+     * ID values are auto-generated in psa_crypto_driver_wrappers.h
+     * ID value zero means the context is not valid or not assigned to
+     * any driver (i.e. none of the driver contexts are active). */
+    unsigned int MBEDTLS_PRIVATE(id);
+    /* Algorithm used for PAKE operation */
+    psa_algorithm_t MBEDTLS_PRIVATE(alg);
+    /* Based on stage (collecting inputs/computation) we select active structure of data union.
+     * While switching stage (when driver setup is called) collected inputs
+       are copied to the corresponding operation context. */
+    uint8_t MBEDTLS_PRIVATE(stage);
+    /* Holds computation stage of the PAKE algorithms. */
+    psa_pake_computation_stage_t MBEDTLS_PRIVATE(computation_stage);
+    union {
+        unsigned dummy;
```

Comment:
You only need a `dummy` member in a union when there are configurations where it would otherwise be empty. Here the union always has two useful members, it doesn't need `dummy`.

Reply:
I removed dummy fields and now I know why I left them. Now I can't make struct initializer macro that will satisfy `gcc` and `clang`.

Current CI failure for `clang`:
```

[2023-01-18T18:35:47.042Z] /var/lib/build/include/psa/crypto_extra.h:2067:43: error: missing field 'sequence' initializer [-Werror,-Wmissing-field-initializers]
[2023-01-18T18:35:47.042Z]     const struct psa_pake_operation_s v = PSA_PAKE_OPERATION_INIT;
[2023-01-18T18:35:47.042Z]                                           ^
[2023-01-18T18:35:47.042Z] /var/lib/build/include/psa/crypto_extra.h:1900:43: note: expanded from macro 'PSA_PAKE_OPERATION_INIT'
[2023-01-18T18:35:47.042Z]                                   { { { 0 } } }, { { 0 } } }
[2023-01-18T18:35:47.042Z]                                           ^
[2023-01-18T18:35:47.042Z] /var/lib/build/include/psa/crypto_extra.h:2067:43: error: missing field 'password_len' initializer [-Werror,-Wmissing-field-initializers]
[2023-01-18T18:35:47.042Z] /var/lib/build/include/psa/crypto_extra.h:1900:56: note: expanded from macro 'PSA_PAKE_OPERATION_INIT'
[2023-01-18T18:35:47.042Z]                                   { { { 0 } } }, { { 0 } } }
[2023-01-18T18:35:47.042Z]                                                        ^
[2023-01-18T18:35:47.300Z] 2 errors generated.
```

Reply:
You need enough `, 0` in the braces to explicitly initialize all the members of the struct that is the first member in the union. (Otherwise it's perfectly valid C with the desired semantics, but `clang -Wall -Wextra` isn't happy.)